### DISCUSSION TITLE
rdev breaks the app on key press

### DISCUSSION THREAD

**Discussion by HuakunShen** [tauri-apps/tauri#7839] [Q&A] [answered]
I read this discussion but my problem is different. https://github.com/tauri-apps/tauri/discussions/7752

I am using MacOS and rdev 0.5.3, trying to listen to key press events.

If I run the listen function in `tauri::async_runtime::spawn(async move ...`, only mouse move events are printed out. 
Whenever I press a key, the program will exit (exit code is 0 by `echo $?`).

If I run the listener without async runtime, which will block setup function, the listener works for key presses and won't exit, but the tauri app is not usable as setup() never finishes.

**Comment by HuakunShen** [comment]
I tried the same code on windows and Ubuntu 20.04, everything works fine. I also tried the code in a fresh MacOS VM but it quits whenever a key is pressed.

Pretty sure it's some problem only on Mac. 
Any suggestion?

**Comment by zengkaiwen** [comment]
I have the same problem 😣，have you solved it?

**Reply by HuakunShen** [reply]
Not yet, no clue where the bug is from, there is no error message.

**Comment by NotAJocke** [comment]
Same problem here, macos and rdev too

**Comment by helbing** [comment]
Same issue in macOS Sonoma. I found that [rustdesk](https://github.com/rustdesk/rustdesk/blob/master/Cargo.toml) use this repo to [solve](https://github.com/rustdesk/rustdesk/blob/master/Cargo.toml) it, and it can work.

**Reply by HuakunShen** [reply, marked as answer]
> Same issue in macOS Sonoma. I found that [rustdesk](https://github.com/rustdesk/rustdesk/blob/master/Cargo.toml) use this repo to [solve](https://github.com/rustdesk/rustdesk/blob/master/Cargo.toml) it, and it can work.

@helbing Thanks!

I tried this it works on Mac.

```toml
rdev = { git = "https://github.com/fufesou/rdev" }
```

If I use a tag from https://github.com/fufesou/rdev/tags 0.3.6 (latest tag in 2020), it still fails. Make sure to use the git url directly.

**Reply by AlexKnauth** [reply]
Do you have any idea what code change or which commit in the https://github.com/fufesou/rdev or https://github.com/rustdesk-org/rdev repository fixed this issue?

**Reply by HuakunShen** [reply]
@AlexKnauth No, to find the commit I am afraid you have to do a binary search on the commit history. I haven't done it.

**Reply by AlexKnauth** [reply]
I found where the crash was coming from, and I think I've found a fix: https://github.com/Narsil/rdev/pull/147, inspired by what https://github.com/fufesou/rdev and https://github.com/rustdesk-org/rdev have done.

**Reply by HuakunShen** [reply]
Cool! What did you fix?

**Reply by Madd0g** [reply]
does anyone have a github link to a tauri app that uses rdev correctly?

I'm new to rust and tauri and the whole requirement about running on the main thread confuses me.

Thanks

**Reply by HuakunShen** [reply]
@Madd0g You can take a look at this plugin
https://github.com/kunkunsh/tauri-plugin-user-input
I haven't documented it

I am using it in https://github.com/kunkunsh/kunkun

**Reply by Madd0g** [reply]
@HuakunShen - absolutely fantastic stuff! thank you!

you got great projects!

**Comment by sincekmori** [comment]
I hit this same crash and this thread was one of the few places discussing it, so let me pay back with what I've learned since.

The root cause was eventually tracked down by @AlexKnauth in Narsil/rdev#147: calling `listen` off the main thread needs `set_is_main_thread(false)`. So the `fufesou/rdev` git workaround isn't necessary anymore, though the fix still isn't released on crates.io, so you need a git dependency on `Narsil/rdev` main instead.

Also, if the reason you're listening for keys is a copy or clipboard workflow (it was for me), there's a way out that avoids rdev on macOS entirely: a listen-only `CGEventTap` on the main run loop, with the main-thread-only pasteboard and window calls hopped over via libdispatch. I ended up packaging that as [copycopy](https://github.com/sincekmori/copycopy) after debugging this. It starts from Tauri's `setup()` and hands you clipboard content plus foreground app info on a worker thread.
